test/unit_algodiff_ad.ml

```diff
@@ -44,9 +45,358 @@ module To_test = struct
 
   let poly9 x = x |> diff (diff (diff pfunc)) |> unpack_flt |> approx_equal 0.00636573616030225702
 
+  let derivative_delta f df x = ((diff f (F x)) |> unpack_flt) -. (df x)
+  (** derivative_delta f df x returns the delta between the *algodiff*
+     derivative of f at x and the function df applied to x
+
+     f : a Maths function D.t -> D.t which maps a D.t 'x' value (always an (F
+     float) here) to f(x)
+
+     df : float -> float maps x to the value of df/dx at x
+
+     x : float **)
+
+  let check_derivative ?eps:(eps=0.) f df x =
+    let delta = derivative_delta f df x in
+    (abs_float delta) < eps
+  (** check_derivative ?eps f df x returns true if the absolute delta between
+     the *algodiff* derivative calculation and the provided df at x is smaller
+     than epsilon
+
+     f : a Maths function D.t -> D.t which maps a D.t 'x' value (always an (F
+     float) here) to f(x)
+
+     df : float -> float maps x to the value of df/dx at x
+
+     x : float **)
+
+  let check_derivative_array ?eps:(eps=0.) f df x_arr =
+    M.print x_arr;
+    let df_arr   = M.mapi (fun o x -> df x) x_arr in
+    M.print df_arr;
+    let diff_arr = (diff f (Arr x_arr)) |> unpack_arr in
```

Comment:
This line should be 

```
let diff_arr = M.map (fun x -> diff f (F x) |> unpack_flt) x_arr in
```

Reply:
A further note on this: 

The reason is: if you pass in a vector to a function, based on your output (scalar or vector), it can be either jacobian or hessian if you want to differentiate a function, so it will confuse `Algodiff` if you don’t specify clearly.